### DISCUSSION TITLE
Multi Client / Multi CA Approach

### DISCUSSION THREAD

**Discussion by irenaeusbecker** [grindsa/acme2certifier#116] [Ideas]
Hi all,

just found your ACME2Certifier based on internet search which looks very nice for my potential project to support multiple CA's etc. 

Just an idea to discuss... is it possible to extend your project to support a workflow and order process which orders a certificate in the background at different ca's and potentially will hand over the certificate back to ACME Server? The idea is to support multiple CAs for one customer with one ACME Server.

Best would be also to support multiple customers on one platform by using templates, but this is just a nice to have.

Would be great to start a discussion with you. 

![Clipboard01](https://github.com/grindsa/acme2certifier/assets/44123946/25f9e2e8-cbeb-4da6-89ce-4662156a21cb)

**Comment by grindsa** [comment]
Hi, 

the idea is nice but requires significant changes on the a2c architecture.  I am not sure if it is worth the effort as i belive that a solution with multiple a2c instances on one host and a proxy like Traefik, Caddy or NGINX in front would do the same job. Customer provision could be done via an external system and then provide eab-credentils to client which can be used for authentication towards the respective a2c instance.

Thoughts?

**Comment by irenaeusbecker** [comment]
Hi,

i agree to your approach in the front by using proxy etc which can be done with less effort. 
Can you please explain a little bit more the handover of requested CA certificates back to acme2certifier so that customer will finally get his requested certificate? EAB Credentials will surely help to have a unique identifier but is there today a potential API to connect such kind of external system? 

Thanks in advance!

**Comment by grindsa** [comment]
Hi,

One idea is to have one a2c instance per CA. All instances are reachable via different endpoints (`https://a2c.foo/ca1/acme/directory`, `https://a2c.foo/ca2/acme/directory`). Every instance would have its own set of eab-credentials. In such scenario customers would share instances. By reaching to a certain instance, they would get the right certificate. This is already possible as of today.

Another option would be to support multiple CAs per instances.  CA selection could be done based on eab-credentials meaning you have a 1:1 mapping between eab-account and CA. This is not fully according to the standard but well... its also not against the standard.  This would come with some code-changes, but the effort should be reasonable. 

The eab-credential management in both cases could be done in an external CRM system.  a2c would connect to this system by using a customized [EAB handler](https://github.com/grindsa/acme2certifier/wiki/External-Account-Binding). Identification and authentication of an a2c instance towards the CRM system can be done by configuration parameters to be inserted in `acme_srv.cfg`. 

Hope my explanations make sense. If not let me know...

**Comment by grindsa** [comment]
Just to keep you updated. Starting from [v0.34](https://github.com/grindsa/acme2certifier/releases/tag/0.34) we are introducing a [profiling feature](https://github.com/grindsa/acme2certifier/blob/master/docs/eab_profiling.md) which allows the definition of individual enrollment parameters per acme-user. So assuming that you use "external account binding" you could define different CAs or enrollment profiles per customer. Input format is described in the [a2c documentation](https://github.com/grindsa/acme2certifier/blob/master/docs/eab_profiling.md); provisioning could be done out of the above-mentioned order-database.

**Reply by ctrought** [reply]
@grindsa I am looking for some guidance on how one would implement multi ca via profiles for both acme and non-acme CAs. In my case I am looking at the below 3

1. LetsEncrypt (ACME CA Handler)
2. [Google Trust Services](https://cloud.google.com/certificate-manager/docs/public-ca-tutorial) (ACME CA Handler) 
3. Microsoft Web Enrollment Services (mscertsrv Handler)

I would like the client to be able to select the CA used by passing the profile somehow. I am also interested in having the CAs either open to all clients, with the option of specific CAs being restricted to a subset of clients (via EAB I guess).

I understand in terms of profiling type features the below currently exist but I'm trying to wrap my head around which features would support my goal of offering multi CA without specifying every clients EAB, but also optionally having one or more CAs restricted to certain clients, and whether multi CA with different handlers in a single instance is possible.

**1. EAB profiling**
Acme: https://github.com/grindsa/acme2certifier/blob/master/docs/eab_profiling.md
Mscertsrv: https://github.com/grindsa/acme2certifier/blob/master/docs/mscertsrv.md#eab-profiling

It's not clear to me exactly how this is meant to be used, I think the part I am missing is who provides the kid and hmac. It says it comes from the CA operator

> To enable EAB, the Certificate Authority (CA) operator must provide both the ACME client and acme2certifier with a Key Identifier (kid) and a MAC key (mac_key). These credentials authenticate NewAccount requests.

which I am assuming means the CAs sitting behind a2c (LetsEncrypt, GoogleTrustServices, mscertsrv in my case). But mscertsrv is not an ACME CA and it seems to support EAB profiling, so how is it providing an EAB kid and mac? I think I may be misinterpreting what CA operator means in this context. Would I instead be looking at building something completely custom to generate kid and hmacs for clients through some other external tool outside a2c, and not attached to any of the CAs behind a2c (Letsencrypt, Google Trust, and Mscertsrv in my case).

**2. Profile ID Header:**
Acme: https://github.com/grindsa/acme2certifier/blob/master/docs/acme_ca.md#passing-a-profile-id
Mscertsrv: https://github.com/grindsa/acme2certifier/blob/master/docs/mscertsrv.md#passing-a-template-from-client-to-server

This feature might provide the option to specify letsencrypt or googletrust since they are ACME backends, though one issue I could see is if I cannot limit what backend CAs can be used (ie. client can choose another acme provider and put that URL in and AC2 will attempt to request a cert from there even if I only want to expose specific ACME CAs). 

In order to support a non ACME CA I imagine I would need to spin a second instance of AC2 up specifically for that CA and then I could offer that endpoint up as another proxy to it? If this is feasible, I would have 2 acme instances where 1 is the main proxy clients interact with, while a second instance is running to handle the mscertsrv CA requests.

**3. Acme Profile Extension:**
Acme only: https://github.com/grindsa/acme2certifier/blob/master/docs/acme_profiling.md

This feels the most natural since acme clients will support profile selection and it could also allow it to fallback to a default profile if some profiles need to be removed without requiring the client to be reconfigured. However based on my understanding of the docs I'm not clear whether it it proxies the profile itself  to the backend request to the ACME CA, and if so then would not work if that profile doesn't also exist on the backend CA. 

Example:
Profile 1: letsencrypt
Profile 2: googletrust
Profile 3: mscertsrv

Config:
```
[Order]
profiles: {"letsencrypt": "https://acme-v02.api.letsencrypt.org/directory", "googletrust": "https://dv.acme-v02.api.pki.goog/directory", "profile3": "https://second-a2c-instance-for-mscertsrv/directory"}
```

If the profile the client passes lets a2c choose the backend CA, this approach would work but if the profile is always proxied to the backend CA in the second ACME request then it might fail since the CAs won't have these profiles available (or it might just fallback to the default profile but would then mean I can only use the default profile). If that is the case, would a new feature to extend the acme profile extension to be able to make frontend and backend profiles independent, or even make it similar to the EAB profiling feature where the full CAhandler can be customized per profile which would allow things like allowed_domainlist per profile? Even better if both this + EAB profiling could be enabled at the same time to support the use case of restricting some CAs to specific clients, while leaving others completely open.

I realize I've likely misinterpreted some of these features/options, but would appreciate your thoughts on what I am trying to achieve. Thank you!

**Reply by grindsa** [reply]
Hi,

Let me comment on your the points:

**1. EAB Profiling:**
EAB profiling is strictly related to acme2certifier and does not depend on the CAs served by A2C.
You can generate your own kids/MACs and assign CA-handler-specific enrollment parameters per EAB-KID.

For microsoft_ca_handler, you can customize parameters such as _template_, _ca_name_, _api_host_, _username_ or _password_.

For acme_ca_handler, you can customize _api_url_ and _acme_keyfile_

This means you can switch between Let’sEncrypt or GoogleTrust based on the EAB-KID/MAC you generate and assign to your clients.

You cannot switch between ACME and Microsoft CAs. This is on the roadmap, but I don’t have a timeline yet, as it requires significant refactoring.

**2. Profile ID Header:**
I recommend not using it for the reasons you mentioned. Additionally, it’s proprietary. Instead, consider using the ACME profile extension if you have a relevant use case.

**3. ACME Profile Extension:**
The current implementation allows clients to select CA templates only.
Switching the CA is not supported at the moment. Your idea is interesting and worth considering, but again it’s a mid-term roadmap item since it would require major rework.

**Summary for Your Use Case:**

- With the current implementation, you cannot switch out-of-the-box between the three CAs you are using.
- You can switch between the two ACME CAs using the EAB profiling feature.
- _Creative workaround_: Deploy a second A2C instance in front of your Microsoft CA. This would effectively turn your Microsoft CA into an ACME-CA, enabling you to use EAB-profiling to potentially achieve your goal.

/G

**Comment by azlux** [comment]
On my side,
I currently implement my own ca-handler how call others ca_handler following a decision "tree" (It's a very simple if/else switch, I assure you).

**Comment by grindsa** [comment]
Hi,

Small update: Multi-CAhandler support is a feature for v0.46 coming out around end of September.

I have some [early code for testing](https://github.com/grindsa/acme2certifier/tree/feat_multi_cahandler) but it still need lots of polishing and verification.

There is also some [feature documentation](https://github.com/grindsa/acme2certifier/blob/feat_multi_cahandler/docs/multi_cahandler.md) available

Feel free to have a look and provide feedback.

/G
